ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,300 @@
+    # Imputation using median strategy only works with numerical data
+    except ValueError:
+        raise ImputeStrategyNotAvailableError(
+            "Can only impute numerical data using mean strategy. Try to restrict imputation"
+            "to certain columns using replacement parameter or use a different mode."
+        )
+
+    return adata
+
+
+def _median(adata: AnnData, **kwargs) -> AnnData:
+    """
+    Impute AnnData object using median imputation. This works only for numerical data.
+
+    Args:
+        adata: The AnnData object to use median Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use median imputation on (if None -> all columns)
+
+    Returns:
+            The imputed AnnData object
+
+    """
+    replacement = kwargs.get("replacement")
+    imputer = SimpleImputer(strategy="median")
+    try:
+        # impute a subset of columns
+        if isinstance(replacement, list):
+            column_indices = get_column_indices(adata, replacement)
+            adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+        # impute all columns if None passed
+        else:
+            adata.X = imputer.fit_transform(adata.X)
+    # Imputation using median strategy only works with numerical data
+    except ValueError:
+        raise ImputeStrategyNotAvailableError(
+            "Can only impute numerical data using median strategy. Try to restrict imputation"
+            "to certain columns using replacement parameter or use a different mode."
+        )
+
+    return adata
+
+
+def _most_frequent(adata: AnnData, **kwargs) -> AnnData:
+    """
+    Impute AnnData object using most_frequent imputation. This works for both, numerical and non-numerical data.
+
+    Args:
+        adata: The AnnData object to use most_frequent Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use most_frequent imputation on (if None -> all columns)
+
+    Returns:
+            The imputed AnnData object
+
+    """
+    replacement = kwargs.get("replacement")
+    imputer = SimpleImputer(strategy="most_frequent")
+    # impute a subset of columns
+    if isinstance(replacement, list):
+        column_indices = get_column_indices(adata, replacement)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+    return adata
+
+
+# ===================== KNN Imputation =========================
+
+
+def _knn(adata: AnnData, **kwargs) -> AnnData:
+    """Impute data using the KNN-Imputer.
+    When using KNN Imputation with mixed data (non-numerical and numerical), encoding using ordinal encoding is required
+    since KNN Imputation can only work on numerical data. The encoding itself is just a utility and will be redone once
```

Comment:
```suggestion
    since KNN Imputation can only work on numerical data. The encoding itself is just a utility and will be undone once
```

undone and not redone, right?